### PR TITLE
implement Radio Buttons and Checkboxes

### PR DESCRIPTION
**Pull request by Andi1986**
Hello, i have implmented 2 of the missing features, Radio Buttons and Checkboxes
Could you have a look and maybe aprove?

One notice: according to https://api.slack.com/reference/block-kit/interactive-components "Messages" can not have radios or checkboxes, this i have **not** validated anywhere.

### CODE REVIEW

File: src/Type.php [hunk 1]
```diff
@@ -151,10 +151,10 @@ abstract class Type
 
         // Inputs
         Inputs\Button::class          => self::BUTTON,
-        // Inputs\Checkboxes::class   => self::CHECKBOXES, // Not yet supported.
+        Inputs\Checkboxes::class   => self::CHECKBOXES, // Not yet supported.
         Inputs\DatePicker::class      => self::DATEPICKER,
         // Inputs\OverflowMenu::class => self::OVERFLOW_MENU, // Not yet supported.
-        // Inputs\RadioButtons::class => self::RADIO_BUTTONS, // Not yet supported.
+        Inputs\RadioButtons::class => self::RADIO_BUTTONS,
```
Comment (jeremeamia) [inline src/Type.php]: Make sure these `=>` are lined up, please.

File: src/Inputs/RadioButtons.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace Jeremeamia\Slack\BlockKit\Inputs;
+
+use Jeremeamia\Slack\BlockKit\Exception;
+use Jeremeamia\Slack\BlockKit\Partials\Option;
+
+class RadioButtons extends InputElement
+{
+
+    use HasConfirm;
```
Comment (jeremeamia) [inline src/Inputs/RadioButtons.php:13]: Also needs `HasOptions` trait.

File: src/Inputs/RadioButtons.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        $this->options[] = $option;
+
+        if ($isInitial) {
+            $this->initialOption = $option;
+        }
+
+        return $this;
+
+    }
+
```
Comment (jeremeamia) [inline src/Inputs/RadioButtons.php]: You could add this method for setting the `initialOption` (borrowed from an existing element).
```
    /**
     * @param string $name
     * @param string $value
     * @return self
     */
    public function initialOption(string $name, string $value): self
    {
        $this->initialOption = new Option($name, $value);
        $this->initialOption->setParent($this);

        return $this;
    }
```

File: src/Inputs/RadioButtons.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    {
+
+        if (count($this->options) > self::MAX_OPTIONS) {
+            throw new Exception('Option Size cannot exceed %d', [self::MAX_OPTIONS]);
+        }
+
+        if (count($this->options) < self::MIN_OPTIONS) {
+            throw new Exception('Option Size must be at least %d', [self::MIN_OPTIONS]);
+        }
+
```
Comment (jeremeamia) [inline src/Inputs/RadioButtons.php]: You need additional validations:
```php
$this->validateOptions();

if (!empty($this->initialOption)) {
    $this->initialOption->validate();
}

if (!empty($this->confirm)) {
    $this->confirm->validate();
}
```

File: src/Inputs/RadioButtons.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * @return array
+     */
+    public function toArray(): array
+    {
+        $data = parent::toArray();
+
+        if (! empty($this->initialOption)) {
+            $data['initial_option'] = $this->initialOption;
```
Comment (jeremeamia) [inline src/Inputs/RadioButtons.php]: Should be `$this->initialOption->toArray()`.

File: src/Inputs/RadioButtons.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        $data['options'] = $this->options;
```
Comment (jeremeamia) [inline src/Inputs/RadioButtons.php]: Should be `$data += $this->getOptionsAsArray();`

File: src/Inputs/Checkboxes.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace Jeremeamia\Slack\BlockKit\Inputs;
+
+use Jeremeamia\Slack\BlockKit\Exception;
+use Jeremeamia\Slack\BlockKit\Partials\Option;
+
+class Checkboxes extends InputElement
+{
+
+    use HasConfirm;
```
Comment (jeremeamia) [inline src/Inputs/Checkboxes.php:13]: Needs the `HasOptions` trait, too.

File: src/Inputs/RadioButtons.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    private const MIN_OPTIONS = 1;
+    private const MAX_OPTIONS = 10;
+
+    /** @var array|Option[] */
+    private $options = [];
+
+    /** @var Option|null */
+    private $initialOption = null;
+
+    public function addOption(Option $option, bool $isInitial = false): RadioButtons {
```
Comment (jeremeamia) [inline src/Inputs/RadioButtons.php]: You can remove this method and rely on `HasOptions:option()`/`HasOptions::options()` from the trait.

File: src/Inputs/Checkboxes.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    private const MIN_OPTIONS = 1;
+    private const MAX_OPTIONS = 10;
+
+    /** @var array|Option[] */
+    private $options = [];
+
+    /** @var array|Option[] */
+    private $initialOptions = [];
+
+    public function addOption(Option $option, bool $isInitial = false): Checkboxes {
```
Comment (jeremeamia) [inline src/Inputs/Checkboxes.php]: You can remove this method and rely on HasOptions:option()/HasOptions::options() from the trait.

File: src/Inputs/Checkboxes.php [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        $this->options[] = $option;
+
+        if ($isInitial) {
+            $this->initialOptions[] = $option;
+        }
+
+        return $this;
+
+    }
+
```
Comment (jeremeamia) [inline src/Inputs/Checkboxes.php]: You can use this method for setting `initialOptions` (borrowed from an existing element):
```php
    /**
     * @param array $options
     * @return self
     */
    public function initialOptions(array $options): self
    {
        foreach ($options as $name => $value) {
            $option = new Option((string) $name, (string) $value);
            $option->setParent($this);
            $this->initialOptions[] = $option;
        }

        return $this;
    }
```

File: src/Inputs/Checkboxes.php [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    {
+
+        if (count($this->options) > self::MAX_OPTIONS) {
+            throw new Exception('Option Size cannot exceed %d', [self::MAX_OPTIONS]);
+        }
+
+        if (count($this->options) < self::MIN_OPTIONS) {
+            throw new Exception('Option Size must be at least %d', [self::MIN_OPTIONS]);
+        }
+
```
Comment (jeremeamia) [inline src/Inputs/Checkboxes.php]: You need additional validations:
```php
$this->validateOptions();

if (!empty($this->initialOptions)) {
    foreach ($this->initialOptions as $option) {
        $option->validate();
    }
}

if (!empty($this->confirm)) {
    $this->confirm->validate();
}
```

File: src/Inputs/Checkboxes.php [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * @return array
+     */
+    public function toArray(): array
+    {
+        $data = parent::toArray();
+
+        if (! empty($this->initialOptions)) {
+            $data['initial_options'] = $this->initialOptions;
```
Comment (jeremeamia) [inline src/Inputs/Checkboxes.php]: Should be:
```
$data['initial_options'] = array_map(function (Option $option) {
    return $option->toArray();
}, $this->initialOptions);
```

File: src/Inputs/Checkboxes.php [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        $data['options'] = $this->options;
```
Comment (jeremeamia) [inline src/Inputs/Checkboxes.php]: Should be `$data += $this->getOptionsAsArray();`

**Review by jeremeamia — CHANGES_REQUESTED** [review]
Hey, thanks for working on this! There are a few changes I'd like to request, but they are all relatively simple and are aimed and reusing existing code from the repo as well and making some things more consistent with the existing patterns. Mainly it boils down to using the `HasOptions` trait. You will need to update the tests slightly as well afterwards. Finally, please also run [`composer style-fix`](https://github.com/jeremeamia/slack-block-kit/blob/master/composer.json#L40) to take care of any linting issues.

Edit: Also, sorry. My comments came in out of order, so best to read through from the "Files changed" tab.

File: src/Inputs/Checkboxes.php (hunk 12 above)
Comment (Andi1986) [reply src/Inputs/Checkboxes.php]: i thought the conversion will be done implicitly, it actually works without array_map, are you sure it is needed here ?

File: src/Inputs/Checkboxes.php (hunk 11 above)
Comment (Andi1986) [reply src/Inputs/Checkboxes.php]: initial options can only be part of the options array, they are already validated by the options array, still validate them again ?

File: src/Inputs/Checkboxes.php (hunk 12 above)
Comment (jeremeamia) [reply src/Inputs/Checkboxes.php]: It's not necessarily needed for the JSON conversation, but currently every other element's toArray method is recursive, such that toArray's output contains only scalars with the arrays.

File: src/Inputs/Checkboxes.php (hunk 11 above)
Comment (jeremeamia) [reply src/Inputs/Checkboxes.php]: I see what you mean, but please do it for now. Thanks.

**Comment by jeremeamia** [comment]
@Andi1986 You can go ahead and keep the `addOption` methods you have. I actually like having those too:
```
public function addOption(Option $option, bool $isInitial = false): self
{
    $this->options[] = $option;

    if ($isInitial) {
        $this->initialOptions[] = $option;
    }

    return $this;
}
```

I may end up pulling those up into the `HasOptions` trait later, but I'll do that refactoring later.

**Comment by Andi1986** [comment]
> @Andi1986 You can go ahead and keep the `addOption` methods you have. I actually like having those too:
> 
> ```
> public function addOption(Option $option, bool $isInitial = false): self
> {
>     $this->options[] = $option;
> 
>     if ($isInitial) {
>         $this->initialOptions[] = $option;
>     }
> 
>     return $this;
> }
> ```
> 
> I may end up pulling those up into the `HasOptions` trait later, but I'll do that refactoring later.

i rethought it, i definitly think its better to have a bool inside Option Object, representing an initial one or not, instead of having one options array and one initialoptions array. i will commit that idea

**Comment by Andi1986** [comment]
@jeremeamia i pushed the fixes

**Review by jeremeamia — APPROVED** [review]
Looks good. I might do some refactoring on initial options for these and other elements, but that's all on me. Thanks for these nice additions.

**Comment by Andi1986** [comment]
@jeremeamia im gonna implement overflow menu and more, right now. i want this finished :)

**Pull request merged**